src/security/builtin_plugins/authentication/src/authentication.c

```diff
@@ -2290,1203 +1781,871 @@ validate_handshake_reply_token(
+    handshake->shared_secret_handle_impl = ddsrt_malloc(sizeof(DDS_Security_SharedSecretHandleImpl));
+    handshake->shared_secret_handle_impl->shared_secret = shared_secret;
+    handshake->shared_secret_handle_impl->shared_secret_size = shared_secret_length;
+    memcpy(handshake->shared_secret_handle_impl->challenge1, challenge1_ref_for_shared_secret, DDS_SECURITY_AUTHENTICATION_CHALLENGE_SIZE);
+    memcpy(handshake->shared_secret_handle_impl->challenge2, challenge2_ref_for_shared_secret, DDS_SECURITY_AUTHENTICATION_CHALLENGE_SIZE);
+  }
+
+  {
+    /* setup expiry listener */
+    dds_time_t cert_exp = get_certificate_expiry(handshake->relation->remoteIdentity->identityCert);
+    if (cert_exp == DDS_TIME_INVALID)
     {
-        DDS_Security_long shared_secret_length;
-        unsigned char *shared_secret;
-
-        if (!generate_shared_secret(handshake, &shared_secret, &shared_secret_length, ex)) {
-            goto err_openssl;
-        }
-
-        handshake->shared_secret_handle_impl = ddsrt_malloc( sizeof(DDS_Security_SharedSecretHandleImpl));
-        handshake->shared_secret_handle_impl->shared_secret = shared_secret;
-        handshake->shared_secret_handle_impl->shared_secret_size = shared_secret_length;
-
-        /* put references to challenge1 and challenge2 into shared secret object */
-        memcpy( handshake->shared_secret_handle_impl->challenge1, challenge1_ref_for_shared_secret, DDS_SECURITY_AUTHENTICATION_CHALLENGE_SIZE);
-        memcpy( handshake->shared_secret_handle_impl->challenge2, challenge2_ref_for_shared_secret, DDS_SECURITY_AUTHENTICATION_CHALLENGE_SIZE);
+      DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, DDS_SECURITY_VALIDATION_FAILED, "Expiry date of the certificate is invalid");
+      goto err_invalid_expiry;
     }
-
-    { /* setup expiry listener */
-        dds_time_t certExpiry = get_certificate_expiry( handshake->relation->remoteIdentity->identityCert );
-
-        if( certExpiry == DDS_TIME_INVALID ){
-          DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, (int)result, "Expiry date of the certificate is invalid");
-          goto err_invalid_expiry;
-        } else if( certExpiry != DDS_NEVER ){
-            add_validity_end_trigger( impl,
-                            IDENTITY_HANDLE( handshake->relation->remoteIdentity ),
-                            certExpiry);
-        }
-
-    }
-
-    ddsrt_mutex_unlock(&impl->lock);
-
-    return result;
+    else if (cert_exp != DDS_NEVER)
+      add_validity_end_trigger(impl, IDENTITY_HANDLE(handshake->relation->remoteIdentity), cert_exp);
+  }
+  ddsrt_mutex_unlock(&impl->lock);
+  return result;
 
 err_invalid_expiry:
-    ddsrt_free( handshake->shared_secret_handle_impl->shared_secret );
-    ddsrt_free( handshake->shared_secret_handle_impl );
+  ddsrt_free(handshake->shared_secret_handle_impl->shared_secret);
+  ddsrt_free(handshake->shared_secret_handle_impl);
```

Comment:
This leaves a dangling pointer in ``handshake->shared_secret_handle_impl``, and IINM can lead to a double free